streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -870,14 +904,18 @@ public void onAssignment(final Assignment assignment) {
                     usedSubscriptionMetadataVersion = leaderSupportedVersion;
                 }
                 processVersionFourAssignment(info, partitions, activeTasks, topicToPartitionInfo);
-                partitionsByHost = info.partitionsByHost();
+                partitionsByHostState = Collections.emptyMap();
+                tasksByHosts = info.tasksByHost();
                 break;
             default:
                 throw new IllegalStateException("This code should never be reached. Please file a bug report at https://issues.apache.org/jira/projects/KAFKA/");
         }
 
         taskManager.setClusterMetadata(Cluster.empty().withPartitions(topicToPartitionInfo));
-        taskManager.setPartitionsByHostState(partitionsByHost);
+        if (receivedAssignmentMetadataVersion >= 4)
+            taskManager.setTasksByHostState(tasksByHosts, receivedAssignmentMetadataVersion);
+        else
+            taskManager.setPartitionsByHostState(partitionsByHostState, receivedAssignmentMetadataVersion);
```

Comment:
I am wondering, if we could unify `setTasksByHostState` and `setPartitionsByHostState` into one method? If we can translate `tasksByHosts -> partitionsByHostState` (or reverse) we could simplify the code (ie, translate once after rebalance, and have unified code everywhere else?)

Reply:
I can remove partitionsByHostState altogether from AssignmentInfo and convert it into tasksByHosts for older versions as well in the assign function. So, all the assignmentInfo's be it from older version or newer will have tasksByHosts. But with this should we update the version of AssingmentInfo to 4 or keep it whatever it is originally.